File: JS/Linked Lists/DoubleLinkedList.js
/*
 * --------------------------------------------------------------------------------
 * File :         DoubleLinkedList.js
 * Project :      DS
 * Author :       Saurish Phatak
 *
 *
 * Description : Double Linked List in JavaScript
 * --------------------------------------------------------------------------------
 *
 * Revision History :
 * 2020-August-01	[SP] : Created
 * --------------------------------------------------------------------------------
 * Copyright (c) 2020 Saurish Phatak
 *
 * License :
 *
 *    This code is meant for learning.
 *    Do NOT Copy Paste it as it may not help in understanding the code.
 *    Understand and type it yourself.
 *
 * DISCLAIMER :
 *    This code may contain unintentional and intentional bugs.
 *    There are no guarantees of this code working properly.
 * --------------------------------------------------------------------------------
 */

// Represents a value in the Double Linked List
function ListNode(value) {
    // Holds the value of the Node
    this.value = value;

    // Pointers to the next and previous nodes in the List
    this.previous = this.next = null;
}

// Represents a Double Linked List
class DoubleLinkedList {
    // Pointers to the head and tail of the List
    #head;
    #tail;

    constructor() {
        // Initially the List is empty
        this.#head = this.#tail = null;
    }

    // Adds a value to the back of the List
    pushBack(value) {
        // Create a new Node
        let newNode = new ListNode(value);

        // If the head and tail are empty
        if (!this.#head && !this.#tail)
            // This is the first Node in the List
            this.#head = newNode;

        // There are already values in the List
        // Add the newNode at the tail
        else {
            // Link the tail and the newNode with each other
            this.#tail.next = newNode;
            newNode.previous = this.#tail;
        }

        // Point the tail to newNode
        this.#tail = newNode;
    }

    // Removes value from the List
    remove(value) {
        // Search for the value in the List
        for (let current = this.#head; current; current = current.next) {
            // Value found
            if (value == current.value) {
                // If this is the ONLY value in the List
                if (this.#head == this.#tail) {
                    // Set head and tail to null
                    this.#head = this.#tail = null;

                    return true;
                }

                // If this value is at the tail
                if (current == this.#tail) {
                    // Set tail to its previous and make its next hull
                    this.#tail = this.#tail.previous;
                    this.#tail.next = null;

                    return true;
                }

                // If this value is at the head
                if (current == this.#head) {
                    // Make the head jump ahead
                    this.#head = this.#head.next;

                    // Disconnect current entirely from the List
                    this.#head.previous = null;
                    current.next = null;

                    return true;
                }

                // The value is in the middle of the List
                // Make the current's surrounding nodes 
                // point to each other
                current.next.previous = current.previous;
                current.previous.next = current.next;

                return true;
            }
        }

        // Value not found
        return false;
    }

    // Print the List forward
    printForward() {
        let output = "";
        for (let current = this.#head; current; current = current.next) {
            output += current.value;

            if (current.next)
                output += " <-> ";
        }

        console.log(output);
    }

    // Print the List backward
    printBackward() {
        let output = "";
        for (let current = this.#tail; current; current = current.previous) {
            output += current.value;

            if (current.previous)
                output += " <-> ";
        }

        console.log(output);
    }

    // Clear the List
    removeAll() {
        // Start from the head and delete each node upto tail
        while (this.#head != this.#tail) {
            // Move the head forward and disconnect it from 
            // the previous node
            this.#head = this.#head.next;
            this.#head.previous.next = null;
            this.#head.previous = null;
        }

        // Set head and tail to null
        this.#head = this.#tail = null;
    }
}

// Create a new List
let list = new DoubleLinkedList();

list.pushBack(5);
list.pushBack(7);
list.pushBack(1);

list.printForward();

list.remove(5);
list.printBackward();


list.remove(7);
list.printForward();


console.log("Done");

File: CPP/Linked Lists/DoubleLinkedList.cpp
/*
 * --------------------------------------------------------------------------------
 * File :         DoubleLinkedList.cpp
 * Project :      CPP
 * Author :       Saurish Phatak
 *
 *
 * Description : Double Linked List in C++
 * --------------------------------------------------------------------------------
 *
 * Revision History :
 * 2020-August-12	[SP] : Added clear() and tweaked printForward()
 * 2020-August-01	[SP] : Created
 * --------------------------------------------------------------------------------
 */

#include <iostream>

using namespace std;

// Node represents a value in the List
struct Node
{
    // Points to the previous Node
    Node *previous;

    // Holds the value of the Node
    int value;

    // Points to the next Node
    Node *next;

    // Constructor
    Node(int v)
    {
        // Initially the previous and next
        // don't point to anything
        this->previous = nullptr;
        this->value = v;
        this->next = nullptr;
    }
};
// Represents a Double Linked List
class DoubleLinkedList
{
    // Points to the Head
    Node *head;

    // Points to the Tail
    Node *tail;

public:
    // Default Constructor
    DoubleLinkedList()
    {
        // Head and Tail are initially nullptr
        // as the List is empty
        head = tail = nullptr;
    }

    // Method to add a Node at the Back of the List
    void pushBack(int value)
    {
        // Holds the new Node
        Node *newNode;

        // Allocate memory for the new Node
        if (nullptr == (newNode = new Node(value)))
            return;

        // If this is the First Node in the List
        if (head == nullptr && tail == nullptr)
        {
            // Make the head point to the newNode
            head = newNode;
        }

        // If the List already contains Nodes
        else
        {
            // Point the newNode's previous to tail's previous
            newNode->previous = tail;

            // Connect newNode's previous Node to newNode
            newNode->previous->next = newNode;
        }

        // Point the tail to newNode
        tail = newNode;
    }

    // Method to remove a Node from the List
    bool remove(int valueToRemove)
    {
        // Search for the value
        for (Node *current = head; current; current = current->next)
        {
            // Value found!
            if (current->value == valueToRemove)
            {
                // If this is the Only Node in List
                if (head == current && tail == current)
                {
                    // Delete the current Node
                    // set head and tail to nullptr
                    delete current;
                    head = nullptr;
                    tail = nullptr;
                    return true;
                }

                // If the value is at the tail
                if (current == tail)
                {
                    // Make the tail point to current's previous
                    tail = current->previous;

                    // Set tail's next to nullptr
                    tail->next = nullptr;

                    // Delete the current Node
                    delete current;
                    return true;
                }

                // If the value is at the head
                if (current == head)
                {
                    // Make the head point to current's next
                    head = current->next;

                    // Set the head's previous to nullptr
                    head->previous = nullptr;

                    // Delete the current Node
                    delete current;
                    return true;
                }

                // The value is in the middle
                // Point the current's previous Node to it's next
                current->previous->next = current->next;

                // Point the current's next Node to it's previous
                current->next->previous = current->previous;
                return true;
            }
        }

        // Value not in the List
        return false;
    }

    // Method to print the List in Forward Direction
    void printForward()
    {
        // Start from the head and go upto tail
        for (Node *current = head; current; current = current->next)
        {
            // Print the current Node's value
            cout << current->value;

            if (current->next)
                cout << " <=> ";
        }
        cout << endl;
    }

    // Method to clear the entire List
    int clear()
    {
        int counter = 0;

        // If the List exists
        if (head && tail)
        {
            // Start from the head and keep on deleting the Nodes
            // until we reach the tail
            while (head != tail)
            {
                // Move the head forward
                head = head->next;

                // Delete head's previous and
                // set it to null
                delete head->previous;
                head->previous = nullptr;
                counter++;
            }

            // Delete the tail
            delete tail;
            counter++;
            head = tail = nullptr;
        }

        return counter;
    }
};

int main()
{
    // Create a new Double Linked List
    DoubleLinkedList list;

    // Add some values to the List
    int value;
    while (cout << "Enter value to add to the List (0 to stop) : ",
           cin >> value,
           value)
    {
        list.pushBack(value);
        list.printForward();
    }

    while (cout << "Enter value to remove from the List (0 to stop) : ",
           cin >> value,
           value)
    {
        list.remove(value);
        list.printForward();
    }

    cout << list.clear() << endl;
    list.printForward();

    return 0;
}